app/src/full/java/io/homeassistant/companion/android/sensors/CarInfoSensorManager.kt

```diff
@@ -0,0 +1,116 @@
+
+    companion object {
+        private const val TAG = "CarInfoSM"
+        private var isListenerRegistered = false
+        private var listenerLastRegistered = 0
+        private val fuelLevel = SensorManager.BasicSensor(
+            "car_info_fuel_level",
+            "sensor",
+            R.string.basic_sensor_name_car_info_fuel_level,
+            R.string.sensor_description_car_info_fuel_level,
+            "mdi:barrel",
+            unitOfMeasurement = "%",
+            stateClass = SensorManager.STATE_CLASS_MEASUREMENT,
+        )
+        private val batteryLevel = SensorManager.BasicSensor(
+            "car_info_battery_level",
+            "sensor",
+            R.string.basic_sensor_name_car_info_battery_level,
+            R.string.sensor_description_car_info_battery_level,
+            "mdi:car-battery",
+            unitOfMeasurement = "%",
+            stateClass = SensorManager.STATE_CLASS_MEASUREMENT,
+        )
+    }
+    private lateinit var latestContext: Context
+
+    override fun docsLink(): String {
+        return "https://companion.home-assistant.io/docs/core/sensors"
+    }
+
+    override fun hasSensor(context: Context): Boolean {
+        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.P
+    }
+
+    override val name: Int
+        get() = R.string.sensor_name_car_info
+
+    override suspend fun getAvailableSensors(context: Context): List<SensorManager.BasicSensor> {
+        return listOf(
+            fuelLevel, batteryLevel
+        )
+    }
+
+    override fun requiredPermissions(sensorId: String): Array<String> {
+        return arrayOf("com.google.android.gms.permission.CAR_FUEL")
+    }
+
+    @RequiresApi(Build.VERSION_CODES.P)
+    override fun requestSensorUpdate(
+        context: Context
+    ) {
+        latestContext = context
+        updateEnergy()
+    }
+
+    @RequiresApi(Build.VERSION_CODES.P)
+    private fun updateEnergy() {
+        if (!isEnabled(latestContext, fuelLevel) && !isEnabled(latestContext, batteryLevel))
+            return
+
+        val now = System.currentTimeMillis()
+        if (listenerLastRegistered + SensorManager.SENSOR_LISTENER_TIMEOUT < now && isListenerRegistered) {
+            Log.d(TAG, "Re-registering listener as it appears to be stuck")
+            HaCarAppService.carInfo?.removeEnergyLevelListener(this)
+            isListenerRegistered = false
+        }
+
+        if(HaCarAppService.carInfo != null) {
+            HaCarAppService.carInfo?.addEnergyLevelListener(latestContext.mainExecutor, this)
+            Log.d(TAG, "CarInfo sensor listener registered")
+            isListenerRegistered = true
+            listenerLastRegistered = now.toInt()
+        } else {
+            Log.d(TAG, "CarInfo not available")
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    override fun onCarDataAvailable(data: EnergyLevel) {
+        if (data.fuelPercent.status == CarValue.STATUS_SUCCESS) {
```

Comment:
We should make sure to skip the updates for the sensors that are not enabled here. It does not seem like we can avoid any particular update here right? The listener looks to be all or nothing.

Reply:
Yes, both values are returned in the same object.

Maybe we could add a check for each sensor like this:
```kotlin
if(data.fuelPercent.status == CarValue.STATUS_SUCCESS && isEnabled(latestContext, fuelLevel))
```